Closes #20301: Add "Dismiss all" action to notifications dropdown

pheus opened the pull request:
### Fixes: #20301

Add a **“Dismiss all”** action to the bell dropdown that clears all **unread** notifications for the current user via a GET endpoint. The dropdown continues to show up to the most recent 10 unread items, and now includes a clear, i18n’d confirmation message that shows the total number of unread notifications affected.

#### Summary of changes

- **New view:** `NotificationDismissAllView` registered with `@register_model_view(Notification, name='dismiss_all', path='dismiss-all', detail=False)`; deletes **unread** notifications for `request.user`.  
  - HTMX requests: if invoked from the Notifications page, return an **`HX-Redirect`** to repaint the page; otherwise return the existing dropdown partial.  
  - Non‑HTMX requests: redirect to the Notifications page.

- **Dropdown partial context:** Now passes only the top **10** unread notifications plus counts:  
  `notifications = request.user.notifications.unread()[:10]`,  
  `total_count = request.user.notifications.count()`,  
  `unread_count = request.user.notifications.unread().count()`. :contentReference[oaicite:1]{index=1}

- **Template update (`htmx/notifications.html`):**  
  Adds a “Dismiss all” button in the dropdown header when there are unread items, with a **pluralized** confirmation that shows the total unread count (e.g., “Dismiss 37 unread notifications?”). 

- **HTMX utility helpers:**  
  Introduce `htmx_current_url()` and `htmx_maybe_redirect_current_page()` to detect the current page (via `HX-Current-URL`) and issue an `HX-Redirect` when appropriate. Both are now used by the single‑item `dismiss` view and the new `dismiss-all` view to avoid stale rows on the Notifications page. :contentReference[oaicite:3]{index=3}


#### Screenshots

**Screenshots (notifications present)**

<img width="302" height="626" alt="Screenshot 2025-10-22 at 14-10-59 Subscriptions NetBox" src="https://github.com/user-attachments/assets/ab9dc1fc-1cf2-4259-971e-df0c977ff16f" />

<img width="302" height="626" alt="Screenshot 2025-10-22 at 14-10-41 Subscriptions NetBox" src="https://github.com/user-attachments/assets/70f47a83-1582-4f3a-a82d-0054b2bd5982" />


**Screenshots (no notifications)**

<img width="302" height="78" alt="Screenshot 2025-10-17 at 00-36-58 Home NetBox" src="https://github.com/user-attachments/assets/98cf4705-20c0-44e5-a9e1-3daf836a1cd8" />

<img width="302" height="78" alt="Screenshot 2025-10-17 at 00-37-04 Home NetBox" src="https://github.com/user-attachments/assets/ad5bf3cf-cc37-418e-b411-9fde9cb46952" />

---

#### Notes for reviewers

- The button label uses **“Dismiss all”** to align with the route name (`dismiss-all`).

bctiemann reviewed (changes requested):
@pheus This works well and after a lot of thought about whether "Clear" was the right terminology (as opposed to "Mark Read" or "Dismiss"), I think probably "Clear" is clear enough, though maybe "Dismiss" is a little more consistent. The issue I have is that our existing terminology isn't very consistent — we use "Delete Selected" for the bulk delete button, and "dismiss" for the URL path, and then we muddle things up with the "read" path for when you click on the notification to view the object it pertains to, which makes it no longer "new" (a dirty badge), but doesn't remove it from your list of notifications, but also doesn't have any differentiator between it and an "unread" notification (well, it's shown in a slightly dimmed gray). "Clear" is a new terminology that has no precedent, but then most other terminology we have is implicit at best. So I'm okay with "Clear" because it obviously does "the right thing" as far as what the user is most likely to want.

I did notice one funny corner case: if you are _viewing_ the Notifications page and you use the "Clear All" button to dismiss all unread notifications, those notifications stay in the list (because the action was done through HTMX and does not repaint the page). Then if you click on the trash can icon for any of the now-deleted notifications, or use the "Delete Selected" button, you get a 404 error.

Could you please give that some thought and see if there's a simple way to avoid that behavior, maybe by forcing a refresh if you're viewing the Notifications page? Or maybe the right answer is it isn't worth trying to prevent that, depending on how much effort it involves.

Thanks!

pheus commented:
Thanks for the thoughtful review!

**Terminology**  
I’ll think this through, but your points have me leaning toward **“Dismiss all”** to match the existing `dismiss-all` route and view. “Clear all” came from the issue text, but I agree we shouldn’t introduce another term.

**Corner case**  
I’ve seen the same behavior when using the ❌ in the dropdown while the notifications table is open. I’ll think it through and report back; if it turns out non‑trivial or brittle, I’m comfortable keeping the current behavior.

Thanks again!

pheus committed:
feat(extras): Add "Dismiss all" action to notifications dropdown

Introduce a view to allow users to dismiss all unread notifications with
a single action. Update the notifications' template to include a
"Dismiss all" button for enhanced usability. This addition streamlines
notification management and improves the user experience.

Fixes #20301

pheus commented on netbox/extras/views.py:
```diff
@@ -4,7 +4,7 @@
 from django.db.models import Count, Q
-from django.http import HttpResponseBadRequest, HttpResponseForbidden, HttpResponse
+from django.http import HttpResponseBadRequest, HttpResponseForbidden, HttpResponse, Http404
 from django.shortcuts import get_object_or_404, redirect, render
```
_Note_: This is an unrelated fix. `Http404` is used, but not imported.
